python/demo/gsoc2022/demo_scattering_boundary_conditions.py

```diff
@@ -0,0 +1,574 @@
+# \bar{\mathbf{v}})-\nabla \times \mathbf{E}_s \cdot \nabla
+# \times\bar{\mathbf{v}}+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s
+# \cdot \bar{\mathbf{v}}+k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_b\right)
+# \mathbf{E}_b \cdot \bar{\mathbf{v}}~\mathrm{dx} \\ +&\int_{\partial \Omega}
+# (\mathbf{n} \times \nabla \times \mathbf{E}_s) \cdot \bar{\mathbf{v}}
+# +\left(j n_bk_{0}+\frac{1}{2r}\right) (\mathbf{n} \times \mathbf{E}_s
+# \times \mathbf{n}) \cdot \bar{\mathbf{v}}~\mathrm{d}s=0,
+# \end{align}
+# $$
+#
+# using the divergence theorem $\int_\Omega\nabla\cdot\mathbf{F}~\mathrm{d}x =
+# \int_{\partial\Omega} \mathbf{F}\cdot\mathbf{n}~\mathrm{d}s$, we can write:
+#
+# $$
+# \begin{align}
+# & \int_{\Omega}-(\nabla \times \mathbf{E}_s) \cdot (\nabla \times
+# \bar{\mathbf{v}})+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s \cdot
+# \bar{\mathbf{v}}+k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_b\right)
+# \mathbf{E}_b \cdot \bar{\mathbf{v}}~\mathrm{d}x \\ +&\int_{\partial \Omega}
+# -(\nabla\times\mathbf{E}_s \times \bar{\mathbf{v}})\cdot\mathbf{n}
+# + (\mathbf{n} \times \nabla \times \mathbf{E}_s) \cdot \bar{\mathbf{v}}
+# +\left(j n_bk_{0}+\frac{1}{2r}\right) (\mathbf{n} \times \mathbf{E}_s
+# \times \mathbf{n}) \cdot \bar{\mathbf{v}}~\mathrm{d}s=0.
+# \end{align}
+# $$
+#
+# We can cancel $-(\nabla\times\mathbf{E}_s \times \bar{\mathbf{V}})
+# \cdot\mathbf{n}$  and $\mathbf{n} \times \nabla \times \mathbf{E}_s
+# \cdot \bar{\mathbf{V}}$ thanks to the triple product rule $\mathbf{A}
+# \cdot(\mathbf{B} \times \mathbf{C})=\mathbf{B} \cdot(\mathbf{C} \times
+# \mathbf{A})=\mathbf{C} \cdot(\mathbf{A} \times \mathbf{B})$, arriving
+# at the final weak form:
+#
+# $$
+# \begin{align}
+# & \int_{\Omega}-(\nabla \times \mathbf{E}_s) \cdot (\nabla \times
+# \bar{\mathbf{v}})+\varepsilon_{r} k_{0}^{2} \mathbf{E}_s \cdot
+# \bar{\mathbf{v}}+k_{0}^{2}\left(\varepsilon_{r}-\varepsilon_b\right)
+# \mathbf{E}_b \cdot \bar{\mathbf{v}}~\mathrm{d}x \\ +&\int_{\partial \Omega}
+# \left(j n_bk_{0}+\frac{1}{2r}\right)( \mathbf{n} \times \mathbf{E}_s \times
+# \mathbf{n}) \cdot \bar{\mathbf{v}} ~\mathrm{d} s = 0.
+# \end{align}
+# $$
+#
+# We can implement such equation in DOLFINx in the following way:
+
+# +
+# Weak form
+F = - inner(curl(Es), curl(v)) * dDom \
+    + eps * k0 ** 2 * inner(Es, v) * dDom \
+    + k0 ** 2 * (eps - eps_bkg) * inner(Eb, v) * dDom \
+    + (1j * k0 * n_bkg + 1 / (2 * r)) \
+    * inner(cross(Es_3d, n_3d), cross(v_3d, n_3d)) * dsbc
+# -
+
+# We can then split the weak form into its left-hand and right-hand side
+# and solve the problem, by storing the scattered field $\mathbf{E}_s$ in `Eh`:
+
+# +
+# Splitting in left-hand side and right-hand side
+a, L = lhs(F), rhs(F)
+
+problem = fem.petsc.LinearProblem(a, L, bcs=[], petsc_options={
+                                  "ksp_type": "preonly", "pc_type": "lu"})
+Eh = problem.solve()
+# -
+
+# Let's now save the solution in a VTK file. In order to do so,
```

Comment:
Should we add some comments about  N1curl fields and why we interpolation into a DG space?